File: app/src/main/kotlin/me/tomasan7/jecnadesktop/ui/component/IconView.java
package me.tomasan7.jecnadesktop.ui.component;

import io.github.palexdev.materialfx.utils.SwingFXUtils;
import javafx.beans.NamedArg;
import javafx.css.*;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.image.WritableImage;
import javafx.scene.paint.Color;

import java.awt.image.BufferedImage;
import java.awt.image.DataBufferInt;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * An extension of {@link ImageView} class, designed to view icons.
 * This class has new style property {@code -jd-icon-color}, which changes the color of the viewed image at runtime.
 * The images are supposed to be single-color images, with possibly transparent pixels.
 * Is able to cache {@link Image Images} for each color that is used, to save time recoloring the images.
 * This behaviour can be changed with {@code doCache} constructor parameter.
 */
public class IconView extends ImageView
{
	private static final StyleablePropertyFactory<IconView> FACTORY = new StyleablePropertyFactory<>(ImageView.getClassCssMetaData());

	private final StyleableObjectProperty<Color> colorProperty = (StyleableObjectProperty<Color>)
			FACTORY.createStyleableColorProperty(this,
												 "color",
												 "-jd-icon-color",
												 IconView::colorPropertyProperty,
												 Color.BLACK,
												 true);

	private final Map<Color, Image> imageCache = new HashMap<>();

	/**
	 * Whether to cache {@link Image} for each {@link Color} that is ever used.
	 * Caching faster but uses more memory, choose depending on the amount of different used colors and icon's size.
	 */
	private final boolean doCache;

	/**
	 * @param image The icon's image.
	 * @param doCache Whether to cache {@link Image} for each {@link Color} that is ever used.
	 *                   Caching faster but uses more memory, choose depending on the amount of different used colors and icon's size.
	 */
	public IconView (Image image, boolean doCache)
	{
		super(image);
		this.doCache = doCache;

		getStyleClass().add("icon-view");

		colorProperty.addListener((__, oldValue, newValue) -> colorize(newValue));

		/* Colorizing with the initial color. */
		colorize(colorPropertyProperty().getValue());
	}

	/**
	 * @param url the string representing the URL from which to load the image.
	 * @param doCache Whether to cache {@link Image} for each {@link Color} that is ever used.
	 *                   Caching faster but uses more memory, choose depending on the amount of different used colors and icon's size.
	 */
	public IconView (@NamedArg("url") String url, @NamedArg(value = "doCache", defaultValue = "true") boolean doCache)
	{
		this(new Image(url), doCache);
	}

	/**
	 * Same as {@link #IconView(Image, boolean)}, but with {@code doCache = false} as default.
	 * Shorthand for {@code new IconView(image, false)}.
	 */
	public IconView (Image image)
	{
		this(image, true);
	}


	/**
	 * Same as {@link #IconView(String, boolean)}, but with {@code doCache = false} as default.
	 * Shorthand for {@code new IconView(url, false)}.
	 */
	public IconView (@NamedArg("url") String url)
	{
		this(url, true);
	}

	private void colorize (Color newColor)
	{
		/*
		 *	TODO: Make the recoloring deal with the Image instances directly
		 *  instead of converting to BufferedImage and then back to Image; unnecessary computing.
		 */

		if (doCache && imageCache.containsKey(newColor))
		{
			imageProperty().set(imageCache.get(newColor));
			return;
		}

		BufferedImage oldBufferedImage = SwingFXUtils.fromFXImage(imageProperty().get(), null);
		recolorBufferedImage(oldBufferedImage, getColor());
		WritableImage fxImage = SwingFXUtils.toFXImage(oldBufferedImage, null);
		imageProperty().set(fxImage);

		imageCache.put(newColor, fxImage);
	}

	public StyleableProperty<Color> colorPropertyProperty ()
	{
		return colorProperty;
	}

	public boolean getDoCache ()
	{
		return doCache;
	}

	public Color getColor ()
	{
		return colorProperty.get();
	}

	public void setColor (Color color)
	{
		colorProperty.set(color);
	}

	@Override
	public List<CssMetaData<? extends Styleable, ?>> getCssMetaData ()
	{
		return FACTORY.getCssMetaData();
	}

	/**
	 * @return The CssMetaData associated with this class, which may include the
	 * CssMetaData of its superclasses.
	 */
	public static List<CssMetaData<? extends Styleable, ?>> getClassCssMetaData ()
	{
		return FACTORY.getCssMetaData();
	}

	/**
	 * Colors every non-transparent pixel of the image to the passed {@link Color}, keeps the alpha values the same.
	 * @param image The {@link Image} to recolor.
	 * @param color The {@link Color} to recolor with.
	 */
	private static void recolorBufferedImage (BufferedImage image, Color color)
	{
		int[] data = ((DataBufferInt) image.getRaster().getDataBuffer()).getData();

		for (int i = 0; i < data.length; i++)
		{
			int pixel = data[i];
			if (pixel == 0)
				continue;

			byte red = (byte) (color.getRed() * 255);
			byte green = (byte) (color.getGreen() * 255);
			byte blue = (byte) (color.getBlue() * 255);

			byte alpha = (byte) ((pixel >> 24) & 0xFF);

			data[i] = (alpha << 24) | ((red & 255) << 16) | ((green & 255) << 8) | (blue & 255);
		}
	}
}

File: src/main/kotlin/me/tomasan7/jecnadesktop/ui/component/IconView.kt
package me.tomasan7.jecnadesktop.ui.component

import io.github.palexdev.materialfx.utils.SwingFXUtils
import javafx.beans.NamedArg
import javafx.css.*
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.paint.Color
import java.awt.image.BufferedImage
import java.awt.image.DataBufferInt

/**
 * An extension of [ImageView] class, designed to view icons.
 * This class has new style property `-jd-icon-color`, which changes the color of the viewed image at runtime.
 * The images are supposed to be single-color images, with possibly transparent pixels.
 * Is able to cache [Images][Image] for each color that is used, to save time recoloring the images.
 * This behaviour can be changed with `doCache` constructor parameter.
 *
 * @param image The icon's image.
 * @param doCache Whether to cache [Image] for each [Color] that is ever used. Caching faster but uses more memory, choose depending on the amount of different used colors and icon's size.
 */
class IconView(image: Image, private val doCache: Boolean = true) : ImageView(image)
{
    val colorProperty: StyleableObjectProperty<Color> = FACTORY.createStyleableColorProperty(
        /* styleable = */ this,
        /* propertyName = */ "color",
        /* cssProperty = */ "-jd-icon-color",
        /* function = */ { it.colorProperty },
        /* initialValue = */ Color.BLACK,
        /* inherits = */ true) as StyleableObjectProperty<Color>

    /**
     * Shorthand for `colorProperty.set(color) and colorProperty.get()`.
     */
    var color: Color
        get() = colorProperty.get()
        set(color) = colorProperty.set(color)

    private val imageCache: MutableMap<Color, Image> = HashMap()

    init
    {
        styleClass.add("icon-view")
        colorProperty.addListener { _, _, newValue: Color -> colorize(newValue) }

        /* Colorizing with the initial color. */
        colorize(color)
    }

    /**
     * Same as [.IconView], but with `doCache = false` as default.
     * Shorthand for `new IconView(url, false)`.
     *
     * @param url the string representing the URL from which to load the image.
     * @param doCache Whether to cache [Image] for each [Color] that is ever used.
     * Caching faster but uses more memory, choose depending on the amount of different used colors and icon's size.
     */
    constructor(@NamedArg("url") url: String?,
                @NamedArg(value = "doCache", defaultValue = "true") doCache: Boolean = true) : this(Image(url), doCache)

    private fun colorize(newColor: Color)
    {
        /*
		 *	TODO: Make the recoloring deal with the Image instances directly
		 *  instead of converting to BufferedImage and then back to Image; unnecessary computing.
		 */
        if (doCache && imageCache.containsKey(newColor))
        {
            imageProperty().set(imageCache[newColor])
            return
        }

        val oldBufferedImage = SwingFXUtils.fromFXImage(imageProperty().get(), null)
        recolorBufferedImage(oldBufferedImage, color)
        val fxImage = SwingFXUtils.toFXImage(oldBufferedImage, null)
        imageProperty().set(fxImage)

        imageCache[newColor] = fxImage
    }

    override fun getCssMetaData(): List<CssMetaData<out Styleable?, *>>
    {
        return FACTORY.cssMetaData
    }

    companion object
    {
        private val FACTORY = StyleablePropertyFactory<IconView>(getClassCssMetaData())

        /**
         * @return The CssMetaData associated with this class, which may include the
         * CssMetaData of its superclasses.
         */
        /* By JavaFX convention, this method should be provided, but unfortunately due to kotlin-java interpretability, it is not possible. */
        /*@JvmStatic
        fun getClassCssMetaData(): List<CssMetaData<out Styleable?, *>> = FACTORY.cssMetaData*/

        /**
         * Colors every non-transparent pixel of the image to the passed [Color], keeps the alpha values the same.
         * @param image The [Image] to recolor.
         * @param color The [Color] to recolor with.
         */
        private fun recolorBufferedImage(image: BufferedImage, color: Color)
        {
            val data = (image.raster.dataBuffer as DataBufferInt).data

            for (i in data.indices)
            {
                val pixel = data[i]
                if (pixel == 0) continue
                val red = (color.red * 255).toInt().toByte()
                val green = (color.green * 255).toInt().toByte()
                val blue = (color.blue * 255).toInt().toByte()
                val alpha = (pixel shr 24 and 0xFF).toByte()
                data[i] = alpha.toInt() shl 24 or (red.toInt() and 255 shl 16) or (green.toInt() and 255 shl 8) or (blue.toInt() and 255)
            }
        }
    }
}